irohad/main/impl/on_demand_ordering_init.cpp

```diff
@@ -117,9 +118,23 @@ namespace iroha {
       return std::make_shared<ordering::OnDemandOrderingGate>(
           std::move(ordering_service),
           std::move(network_client),
-          notifier.get_observable().map(
-              [](auto commit)
-                  -> ordering::OnDemandOrderingGate::BlockRoundEventType {
+          notifier.get_observable()
+              .lift<iroha::synchronizer::SynchronizationEvent>(
+                  iroha::makeTimeout<iroha::synchronizer::SynchronizationEvent>(
+                      [reject_counter = -1](const auto &commit) mutable {
+                        using namespace std::chrono;
+                        using namespace iroha::synchronizer;
+                        if (commit.sync_outcome
+                            == SynchronizationOutcomeType::kReject) {
+                          return seconds(std::min(reject_counter++, 0));
```

Comment:
`reject_counter` may potentially overflow, please add a check so that it does not go above the maximum value.